### PR TITLE
Reoptimize

### PR DESCRIPTION
**Pull request by jstoppelman**
Addition of reopt.py files into the tools directory of ForceBalance. Requested review of the code setup for the "reopt_script_combined" folder and "reopt_script_separate" directories. The code is essentially the same, except the "reopt_script_separate" contains an engines.py code with separate classes and methods for use by the main code, while the "combined" version contains all of these methods into one script.

### CODE REVIEW

**Comment by jstoppelman** [comment]
Hi Lee-Ping, the Drude study should go with the other pull request I have open, I must have mixed up the files. Maybe it's best to have these separate? I updated openmmio.py to be consistent with the current version in ForceBalance and removed the Drude directory. I also updated the About.txt file.

File: src/molecule.py [hunk 1]
```diff
@@ -32,10 +32,6 @@
 except NameError:
     pass
 
-# Special error which is thrown when TINKER .arc data is detected in a .xyz file
```
Comment (leeping) [inline src/molecule.py:35]: This needs to be reverted, perhaps by merging the master branch into this branch.

File: src/molecule.py [hunk 2]
```diff
@@ -173,9 +169,9 @@ class ActuallyArcError(IOError):
 # charge     = The net charge of the molecule
 # mult       = The spin multiplicity of the molecule
 MetaVariableNames = {'fnm', 'ftype', 'qcrems', 'qctemplate', 'qcerr', 'charge', 'mult', 'bonds', 'topology',
-                     'molecules'}
+                     'molecules', 'psi4template', 'psi4fragn', 'psi4args'}
 # Variable names relevant to quantum calculations explicitly
-QuantumVariableNames = {'qcrems', 'qctemplate', 'charge', 'mult', 'qcsuf', 'qm_ghost', 'qm_bondorder'}
+QuantumVariableNames = {'qcrems', 'qctemplate', 'charge', 'mult', 'qcsuf', 'qm_ghost', 'qm_bondorder', 'psi4template', 'psi4fragn',                         'psi4args'}
```
Comment (leeping) [inline src/molecule.py:174]: There are too many spaces between psi4fragn and psi4args.

File: src/molecule.py [hunk 3]
```diff
@@ -246,48 +242,20 @@ def emit(self, record):
             'Fr','Ra','Ac','Th','Pa','U','Np','Pu','Am','Cm','Bk','Cf','Es','Fm','Md','No','Lr','Rf','Db','Sg','Bh','Hs','Mt']
 
 # Dictionary of atomic masses ; also serves as the list of elements (periodic table)
-#
```
Comment (leeping) [inline src/molecule.py:249]: This needs to be reverted, perhaps by merging the master branch into this branch.

File: src/molecule.py (hunk 1 above)
Comment (leeping) [reply src/molecule.py:35]: (As in, the "red" part needs to be put back and the "green" part needs to be removed.)

File: src/molecule.py [hunk 4]
```diff
@@ -313,23 +281,23 @@ def elem_from_atomname(atomname):
             have_dcdlib = True
             break
     if not have_dcdlib:
-        logger.debug('Note: Cannot import optional dcdlib module to read/write DCD files.\n')
```
Comment (leeping) [inline src/molecule.py:316]: The `logger.debug` messages are newer, so the changes here need to be reverted.

File: src/molecule.py [hunk 5]
```diff
@@ -313,23 +281,23 @@ def elem_from_atomname(atomname):
...
+        logger.info('The dcdlib module cannot be imported (Cannot read/write DCD files)')
 
     #============================#
     #| PDB read/write functions |#
     #============================#
     try:
         from .PDB import *
     except ImportError:
-        logger.debug('Note: Cannot import optional pdb module to read/write PDB files.\n')
+        logger.info('The pdb module cannot be imported (Cannot read/write PDB files)')
```
Comment (leeping) [inline src/molecule.py:292]: The `logger.debug` messages are newer, so the changes here need to be reverted.

File: src/molecule.py [hunk 6]
```diff
@@ -313,23 +281,23 @@ def elem_from_atomname(atomname):
...
 
     #=============================#
     #| Mol2 read/write functions |#
     #=============================#
     try:
         from . import Mol2
     except ImportError:
-        logger.debug('Note: Cannot import optional Mol2 module to read .mol2 files.\n')
+        logger.info('The Mol2 module cannot be imported (Cannot read/write Mol2 files)')
```
Comment (leeping) [inline src/molecule.py:300]: The `logger.debug` messages are newer, so the changes here need to be reverted.

File: src/molecule.py [hunk 7]
```diff
@@ -489,8 +456,8 @@ def AStr(self):
             return ','.join(['%i' % i for i in self.L()])
         def e(self):
             """ Return an array of the elements.  For instance ['H' 'C' 'C' 'H']. """
-            elems = nx.get_node_attributes(self,'e')
```
Comment (leeping) [inline src/molecule.py:492]: The lines highlighted in red are newer, so the changes here need to be reverted, perhaps by merging the master branch into this branch.

File: src/molecule.py [hunk 8]
```diff
@@ -500,7 +467,7 @@ def x(self):
             coors = nx.get_node_attributes(self,'x')
             return np.array([coors[i] for i in self.L()])
 except ImportError:
-    logger.warning("Cannot import optional NetworkX module, topology tools won't work\n.")
```
Comment (leeping) [inline src/molecule.py:503]: The lines highlighted in red are newer, so the changes here need to be reverted, perhaps by merging the master branch into this branch.

File: src/molecule.py [hunk 9]
```diff
@@ -1668,17 +1645,17 @@ def write(self,fnm=None,ftype=None,append=False,selection=None,**kwargs):
     #=====================================#
 
     def center_of_mass(self):
-        totMass = sum([PeriodicTable.get(self.elem[i], 0.0) for i in range(self.na)])
```
Comment (leeping) [inline src/molecule.py:1671]: The lines highlighted in red are newer, so the changes here need to be reverted, perhaps by merging the master branch into this branch.

File: src/molecule.py [hunk 10]
```diff
@@ -1668,17 +1645,17 @@ def write(self,fnm=None,ftype=None,append=False,selection=None,**kwargs):
...
-        return np.array([np.sum([xyz[i,:] * PeriodicTable.get(self.elem[i], 0.0) / totMass for i in range(xyz.shape[0])],axis=0) for xyz in self.xyzs])
+        M = sum([PeriodicTable.get(self.elem[i], 0.0) for i in range(self.na)])
+        return np.array([np.sum([xyz[i,:] * PeriodicTable.get(self.elem[i], 0.0) / M for i in range(xyz.shape[0])],axis=0) for xyz in self.xyzs])
 
     def radius_of_gyration(self):
-        totMass = sum([PeriodicTable[self.elem[i]] for i in range(self.na)])
```
Comment (leeping) [inline src/molecule.py:1675]: The lines highlighted in red are newer, so the changes here need to be reverted, perhaps by merging the master branch into this branch.

File: src/molecule.py [hunk 11]
```diff
@@ -2866,7 +2843,7 @@ def read_xyz(self, fnm, **kwargs):
         """ .xyz files can be TINKER formatted which is why we have the try/except here. """
         try:
             return self.read_xyz0(fnm, **kwargs)
-        except ActuallyArcError:
```
Comment (leeping) [inline src/molecule.py:2869]: The lines highlighted in red are newer, so the changes here need to be reverted, perhaps by merging the master branch into this branch.

File: src/molecule.py [hunk 12]
```diff
@@ -2891,28 +2868,11 @@ def read_xyz0(self, fnm, **kwargs):
             if ln == 0:
                 # Skip blank lines.
                 if len(line.strip()) > 0:
-                    try:
```
Comment (leeping) [inline src/molecule.py:2894]: The lines highlighted in red are newer, so the changes here need to be reverted, perhaps by merging the master branch into this branch.

File: src/molecule.py [hunk 13]
```diff
@@ -2987,11 +2947,7 @@ def read_inpcrd(self, fnm, **kwargs):
             if ln == 0:
                 comms = [line]
             elif ln == 1:
-                # Although is isn't exactly up to spec,
```
Comment (leeping) [inline src/molecule.py:2990]: The lines highlighted in red are newer, so the changes here need to be reverted, perhaps by merging the master branch into this branch.

**Review by leeping — CHANGES_REQUESTED** [review]

**Comment by leeping** [comment]
Closing this PR per discussion with @jstoppelman .

**Pull request closed without merging**